### DISCUSSION TITLE
Custom Priors

### DISCUSSION THREAD

**Discussion by FedericoRaimondi** [google/lightweight_mmm#39] [Q&A] [answered]
Hello,

quite exciting to see the new features!
I have one doubt on priors customization.
Media betas it's clear we can customize the halfnormal sigmas by using the weighted cost.
I see we have now also a way to customize non media priors:
> See detailed explanation on custom priors in our documentation.
> custom_priors = {"intercept": numpyro.distributions.Uniform(1, 5)}

Is there a way (maybe it's planned to be added in the documentation) to change decay and saturation parameters?

**Comment by pabloduque0** [comment, marked as answer]
Hey @FedericoRaimondi ! Thanks!

Yes documentation for that should go out tomorrow (the day after at the latest). Let me know if anything js unclear after you go through it

**Reply by pabloduque0** [reply]
@FedericoRaimondi let me know if anything is unclear on the [updated docs](https://lightweight-mmm.readthedocs.io/en/latest/custom_priors.html)

**Reply by FedericoRaimondi** [reply]
Terrific job! It's really clear and covers everything.
From what I understand it's not possible to customize prior by media variable a part from the betas. Am I right?

**Reply by pabloduque0** [reply]
You are right! For the media coefficients for now you can only alter the `scale` of the default HalfNormal.

There is a couple reasons implementation wise that its a bit tricky to allow a custom distribution for this one, as for the geo models you will have a hierarchical model so: `HalfNormal(scale=HalfNormal(scale=media_prior))` when your chosen distribution has more than one parameter its a bit more complicated. Dont get me wrong is very doable I just need to find the right way so its clear for the user how it will work when you have the nested case.

What are other distribution would you like to actually use for `media prior`? Will be very interesting for me to hear what exactly users are trying to do to find the best way to fit it in.

**Reply by FedericoRaimondi** [reply]
Sorry, I think I didn't explain my point correctly. Media priors (betas) is fine to have a Halfnormal where we can adjust the size by media touchpoint (*TV, OOH, Search, Social*). I cannot think at a use case for which I would use a different distribution.

My point was for other media parameters, such as  `half_max_effective_concentration`, `slope`, or `lag_weight`.  From my understanding, I can easily customize the distribution. But, I think we cannot customize it by touchpoint. If I want to give *TV* a very specific `slope` prior that is different from the one I give to *Social*, how would I approach it?

**Reply by pabloduque0** [reply]
Oh yes, you can definitely do that. Lets say you have 3 media channels and you want them to have a different slope:

```
custom_priors = {"slope": numpyro.distributions.Gamma(concentration=jnp.array([1, 1, 3]), rate=jnp.array([1, 1, 0.5]))
mmm.fit(..., custom_priors=custom_priors, ...)
```
In the docs, for each prior you have the "final shape" and you can pass an array for any of the values of the distribution that is broadcastable to that shape.

In the docs at the bottom you can see an example for the "weekday" prior, same idea, 7 days, you can pass different priors for each day.

For the priors that final shape is `()` this is not possible.
